src/restapi/fIncomingWebhooks.ts

```diff
@@ -55,22 +53,147 @@ export async function fHandleSubcribeNotification(req: Request, res: Response):
-
     res.json({});
 }
 
-function getNotificationMessage(zdUrl: string, ticketID: string, ticketTitle: string, auditEvent: any): string {
+// getNotificationMessage returns text for a post message
+async function getNotificationMessage(zdClient: ZDClient, zdUrl: string, ticketID: string, ticketTitle: string, ticketAudit: any): Promise<string> {
     const ZDTicketPath = Routes.ZD.TicketPathPrefix;
-    const ticketLink = `[#${ticketID}](${zdUrl}${ZDTicketPath}/${ticketID})`;
-    const prefix = `Ticket ${ticketLink}  **Subject:** \`${ticketTitle}\` -- `;
-    switch (auditEvent.type) {
-    case 'Comment':
-        return `${prefix}\`${auditEvent.author_id}\` commented on ticket`;
-    case 'Change':
-        // auditEvent.author not defined for field name change;
-        return `${prefix}\`${auditEvent.field_name}\` changed from \`${auditEvent.previous_value}\` to \`${auditEvent.value}\``;
-
-    default:
-        return `type not found. type = ${auditEvent.type})`;
+    const ticketLink = `${zdUrl}${ZDTicketPath}/${ticketID}`;
+    const prefix = `Ticket [#${ticketID}](${ticketLink})  **Subject: [${ticketTitle}](${ticketLink})**  `;
+
+    const auditEvents = ticketAudit.events;
+    const changeEvents = getEventTypes(auditEvents, 'Change');
+    if (changeEvents.length) {
+        const events = await mapIDsToTextValues(zdClient, changeEvents);
+        return prefix + getChangedEventText(events);
+    }
+
+    const createEvents = getEventTypes(auditEvents, 'Create');
+    if (createEvents.length) {
+        const events = await mapIDsToTextValues(zdClient, createEvents);
+        return '**[NEW TICKET]** ' + prefix + getCreatedEventText(events);
+    }
+    throw new Error('Event type does not contain Change or Create event types');
+}
+
+// mapIDsToTextValues takes an array of events and maps ID values to human
+// readable text
+async function mapIDsToTextValues(zdClient: ZDClient, events: any[]): Promise<any> {
+    const mappedArray: any[] = [];
+    const promiseEvents: any[] = [];
+
+    // build the new mapped event with promises for where needed
+    for (const event of events) {
+        if (event.field_name === 'ticket_form_id') {
+            promiseEvents.push(getFormNames(zdClient, event));
+            continue;
+        }
+        if (event.field_name === 'assignee_id') {
+            promiseEvents.push(getAssigneeNames(zdClient, event));
+            continue;
+        }
+        if (event.field_name === 'requester_id') {
+            promiseEvents.push(getAssigneeNames(zdClient, event));
+            continue;
+        }
+        if (event.field_name === 'group_id') {
+            promiseEvents.push(getGroupNames(zdClient, event));
+            continue;
+        }
+        mappedArray.push(event);
+    }
+
+    // resolve all the Promises
+    await Promise.all(promiseEvents).then((mappedEvents) => {
+        for (const mappedEvent of mappedEvents) {
+            mappedArray.push(mappedEvent);
+        }
+    });
```

Comment:
How many requests can this end up being here? Also, is it possible that `mappedEvent` is null?

Is it possible to grab multiple types of values with one request, to limit the number of requests? Some way of reusing data, to avoid over-fetching?

Reply:
I don't have better solution

Reply:
To better answer your question.  

The max number of requests would be on a changed event with all of the field names.. A request is needed for the current name and the previous, so a max number of requests is ..

* 4 * 2 = 8 requests.  
  * 4 (number of fields than need mapped from id to human readable)
  * 2 (retrieve the current and previous values)

**Is it possible that `mappedEvent` is null?**
Each of the `mappedEvents` in the `promiseEvents` array are already await and resolved.  Also each request in the getXXXName() function is checked for an error.  

If any request fails, the webhook will not send and the failed request message is logged.